services/upgrader/upgrader_linux.go

```diff
@@ -0,0 +1,332 @@
+		upgradePercentage = 0
+		return err
+	}
+	defer os.Remove(sigfilename)
+
+	err = verifySignature(filename, sigfilename, mattermostBuildPublicKey)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = errors.New("Unable to verify downloaded file signature")
+		mlog.Error("Unable to verify downloaded file signature", mlog.Err(err))
+		return err
+	}
+
+	err = extractBinary(executablePath, filename)
+	if err != nil {
+		upgradePercentage = 0
+		upgradeError = err
+		mlog.Error("Unable to extract the binary from the downloaded file", mlog.Err(err))
+		return err
+	}
+	upgradePercentage = 100
+	return nil
+}
+
+func UpgradeToE0Status() (int64, error) {
+	return upgradePercentage, upgradeError
+}
+
+func download(url string, limit int64) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	out, err := ioutil.TempFile("", "*_mattermost.tar.gz")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+
+	counter := &writeCounter{total: resp.ContentLength}
+	_, err = io.Copy(out, io.TeeReader(&io.LimitedReader{R: resp.Body, N: limit}, counter))
+	return out.Name(), err
+}
+
+func getFilePermissionsOrDefault(filename string, def os.FileMode) os.FileMode {
+	file, err := os.Open(filename)
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	fileStats, err := file.Stat()
+	if err != nil {
+		mlog.Error("Unable to get the file permissions", mlog.String("filename", filename), mlog.Err(err))
+		return def
+	}
+	return fileStats.Mode()
+}
+
+func extractBinary(executablePath string, filename string) error {
+	gzipStream, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for {
+		header, err := tarReader.Next()
+
+		if err == io.EOF {
+			return errors.New("Unable to find mattermost binary in the downloaded version")
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if header.Typeflag == tar.TypeReg && header.Name == "mattermost/bin/mattermost" {
+			permissions := getFilePermissionsOrDefault(executablePath, 0755)
+			tmpFile, err := ioutil.TempFile(path.Dir(executablePath), "*")
+			if err != nil {
+				return err
+			}
+			tmpFileName := tmpFile.Name()
+			os.Remove(tmpFileName)
+			err = os.Rename(executablePath, tmpFileName)
+			if err != nil {
+				return err
+			}
+			outFile, err := os.Create(executablePath)
+			if err != nil {
+				err2 := os.Rename(tmpFileName, executablePath)
+				if err2 != nil {
+					mlog.Critical("Unable to restore the executable backup. Restore the executable manually.")
+					return errors.Wrap(err2, "critical error: unable to upgrade the binary or restore the old binary version. Please restore it manually")
```

Comment:
```suggestion
					return errors.Wrap(err2, "Critical error: unable to upgrade the binary or restore the old binary version. Please restore it manually.")
```

Can we add more details or instructions for such a critical error?

Reply:
It is the standard approach to not capitalize error strings on golang.